### ISSUE TITLE
Wrong height for nested table

### ISSUE THREAD

**Issue by tjaderxyz** [koreader/koreader#9963]
* KOReader version: v2022.11
* Device: Cervantes 4

#### Issue

Nested table has wrong height.

Firefox:
![image](https://user-images.githubusercontent.com/81907/209865895-afa8559c-0cce-45f7-969f-e961355fb2a5.png)

KOReader:
![koreader](https://user-images.githubusercontent.com/81907/209865922-337f5b9d-9941-4702-8424-b9d61cb800d1.png)

#### Steps to reproduce

Here's a simple example:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>test</title>
    <style>
      table {
        height: 100%;
        border: 1px solid black;
      }
      td {
        border: 1px dashed black;
      }
    </style>
  </head>
  <body>
    <table class="fora">
      <tr>
        <td>
          <table>
            <tr><td>1</td></tr>
            <tr><td>2</td></tr>
          </table>
        </td>
        <td>
          <table>
            <tr><td>1</td></tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
```

**Comment by poire-z** [comment]
crengine doesn't ensure `height` in `%` (because it's too complicated, and would need some going forth and back, which we currently don't do, which makes things a lot easier everywhere), so your `height: 100%` will not be ensured.

As for the vertical alignment, your right cell is aligned to the top.
https://stackoverflow.com/questions/33487148/default-value-of-vertical-align-for-table-cells

It seems to be our (historical) default with `Style> epub.css` - which, well, might be good enough and make things more readable in some context:
![image](https://user-images.githubusercontent.com/24273478/210083363-36422b6f-2d18-4ba1-ba12-795ea7e50165.png)
But you can get a more conforming rendering by using `Style> html5.css` (which is always a good things to try switching to when you get some odd rendering):
![image](https://user-images.githubusercontent.com/24273478/210083441-2e21bf32-cd31-4179-a40e-e391a6f431ae.png)

**Comment by tjaderxyz** [comment]
Centering vertically doesn't really help for my intended usecase, which is a parallel text with the original text on one column and the translated text on the other column. Having the same height on both tables would make things more or less aligned, even if the original and translation have different lengths and differences in paragraph breaks and other typography.

Feel free to close this issue if it's something that's not going to be implemented.

**Comment by poire-z** [comment]
Ensuring the height would just ensure the top and bottom borders (if any) of the left and right cells would align: it would not ensure the text on both sides span the same height - ie. you right text, if smaller in height, would be either centered aligned, or top aligned, as the right `1` in the last screenshots above - you would still get some blank on its top or its top&bottom.
So, I don't think it would be a solution to your use case (?). The solution (if I get your aim right) would be to play with the line-height for the side with the smallest content.

**Comment by tjaderxyz** [comment]
The tables for original and translated would have one row per paragraph, so it would essentially "vartically justify", so that the first paragraph is at the top, the last paragraph is a the bottom, and whitespace is distributed between paragraphs. It works fine in browsers.

```html
<!DOCTYPE html>
<html>
  <head>
    <title>test</title>
    <style>
      table {
        height: 100%;
        border: 1px solid black;
      }
      td {
        border: 1px dashed black;
      }
    </style>
  </head>
  <body>
    <table>
      <tr>
        <td>
          <table>
            <tr><td>1</td></tr>
            <tr><td>2</td></tr>
            <tr><td>3</td></tr>
            <tr><td>4</td></tr>
            <tr><td>5</td></tr>
          </table>
        </td>
        <td>
          <table>
            <tr><td>1</td></tr>
            <tr><td>2</td></tr>
            <tr><td>3</td></tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
```

![firefox](https://user-images.githubusercontent.com/81907/210098279-3a2012af-36e0-4344-a38a-b574adf9b557.png)

**Comment by tjaderxyz** [comment]
One thing that would be even better than nested tables for what I want to do is flexbox layout support, but I assume that would be even harder to implement.

**Comment by poire-z** [comment]
> The tables for original and translated would have one row per paragraph, so it would essentially "vartically justify", 

I see. Well, I'm afraid we can't give you that :/

> flexbox layout support, but I assume that would be even harder to implement.

Indeed, given the specs and the various combinations it would be very hard to implement.

The only things we may somehow support well (to some degree :) are `float` and `inline-box/table`.

For your use case, the only thing I could come up with, that would somehow give you something with crengine - but it's limited to one para left / one para right - and paragraphs needs to be ordered l/r/l/r/l/r/l/r (not l/l/l/l | r/r/r/r like you did) - would be:
```html
<style>
    div.test {
        border: 1px solid black;
        display: inline-block;
        width: 50%;
        box-sizing: border-box;
        vertical-align: middle;
    }
</style>
<div class=test>
abc<br/>
abc<br/>
abc<br/>
def
</div><!-- no space/newline here !!
--><div class=test>
abc<br/>
def
</div><!-- no space/newline here !!
--><div class=test>
abc<br/>
def
</div><!-- no space/newline here !!
--><div class=test>
abc<br/>
abc<br/>
abc<br/>
def
</div>
```
![image](https://user-images.githubusercontent.com/24273478/210101520-cfd0fcbc-63a5-4e8a-a9cc-1aaf799e8585.png)

It would break pages more properly than your original nested tables solution I think (web browsers don't have to care about that, but crengine would, and with nested tables, if your inner tables are taller than a page, a cut would be done at any place, including in the middle of a line, so getting truncated glyphs (or duplicated lines, not sure if we would truncate or display the overflow).

![image](https://user-images.githubusercontent.com/24273478/210101805-e81e5699-1a3d-4122-9874-4c37434da692.png) vs. ![image](https://user-images.githubusercontent.com/24273478/210101874-5966a8f0-16ca-46cc-8c63-07977d2a09c6.png)

**Comment by tjaderxyz** [comment]
Yeah, that would work, as well as what I have done before, having one table with two columns to keep paragraphs aligned.

I was trying this nested table approach because sometimes the original and translation don't match one-to-one in paragraphs, so I can't just blindly interleave them. I could stuff more than one paragraph in a table row or div when the texts didn't match, but that would require manually reading the entire text in order to align stuff.

Reading the entire text with a translation on the side is what I want to do in the first place, so having to do that before hand kind of defeats the point.

Thanks for the suggestions, and feel free to close this issue if percentage height support is something you consider to be out of scope for crengine.
